target_chains/cosmwasm/tools/src/pipeline.ts

```diff
@@ -0,0 +1,181 @@
+    this.pipelineStore = new PipelineStore(filePath);
+  }
+
+  addStage(stage: Stage) {
+    this.stages.push(stage);
+  }
+
+  private stageExecutorWrapper(executor: StageExecutor) {
+    // We want to wrap the executor provided by the pipeline consumer
+    // In order to wrap the response of the executor in the StageResult
+    // also in this method we inject the `getResultOfPastStage` to the stage executor
+    return async (): Promise<StageResult> => {
+      // method to inject
+      const getResultOfPastStage = <Y>(stageId: string): Y => {
+        let result = this.pipelineStore.getStageState<StageResult<Y>>(stageId);
+
+        // This if condition will execute only if the stage executor is
+        // trying to reading a stage's state with stage id that doesn't exist
+        // past results will all be fulfilled and the pipeline will make sure of that
+        if (
+          result === undefined ||
+          (result !== undefined && result.status === "rejected")
+        ) {
+          throw new Error(
+            `${this.id}: Stage id seems to be invalid: ${stageId}`
+          );
+        }
+        return result.result;
+      };
+      try {
+        // wrapping result
+        const result = await executor(getResultOfPastStage);
+        return {
+          status: "fulfilled",
+          result,
+        };
+      } catch (e) {
+        return {
+          status: "rejected",
+          reason: e,
+        };
+      }
+    };
+  }
+
+  private async processStage(stage: Stage): Promise<boolean> {
+    // Here we will check if there is a past result that has been fulfilled
+    // If yes, we are not going to process any further
+    let currentResult = this.pipelineStore.getStageState<StageResult>(stage.id);
+    if (currentResult !== undefined && currentResult.status === "fulfilled")
+      return true;
+
+    // Else we will process the new stage and store the result
+    const newResult = await this.stageExecutorWrapper(stage.executor)();
+    this.pipelineStore.setStageState(stage.id, newResult);
+
+    if (newResult.status === "fulfilled") return true;
+
+    // Some steps can fail due to some one time errors like API issues
+    // This allows the user to re run this particular stage
+    const rerun =
+      (await readLineAsync(
+        `${this.id}: Some steps of stage: ${stage.id} failed. \n Do you want to rerun? (y) `
+      )) === "y";
+
+    if (rerun) return this.processStage(stage);
+    else return false;
+  }
+
+  async run() {
+    console.log("Running pipeline with id: ", this.id);
+    for (let stage of this.stages) {
+      console.log(`${this.id}: Running stage with id: ${stage.id}`);
+
+      // This method is only going to process stage if all the past ones have been fulfilled
+      let fulfilled = await this.processStage(stage);
+      if (fulfilled === false) break;
+    }
+
+    // store the whole processing locally
+    this.pipelineStore.commit();
```

Comment:
you probably want to do this after processing every `stage`, just in case this code crashes for some reason.